lib/galaxy/datatypes/binary.py

```diff
@@ -1040,15 +1047,20 @@ def _layercountsize(tmp, lennames=0):
                     tmp = anndata_file["uns"]
                     dataset.metadata.uns_layers, dataset.metadata.uns_count, _ = _layercountsize(tmp)
 
+                ## Resolving the problematic shape parameter
                 if 'X' in dataset.metadata.layers_names:
                     # Shape we determine here due to the non-standard representation of 'X' dimensions
                     shape = anndata_file['X'].attrs.get("shape")
                     if shape is not None:
                         dataset.metadata.shape = tuple(shape)
                     elif hasattr(anndata_file['X'], 'shape'):
                         dataset.metadata.shape = tuple(anndata_file['X'].shape)
-                    else:
-                        dataset.metadata.shape = (int(dataset.metadata.obs_size), int(dataset.metadata.var_size))
+                ##
+                elif n_obs >=0 and n_var >= 0:
+                    dataset.metadata.shape = (int(n_obs), int(n_var))
+                else:
+                    log.warning("No shape data could be determined!")
+                    dataset.metadata.shape = tuple(-1,-1)
```

Comment:
```suggestion
                else:
                    dataset.metadata.shape = (int(dataset.metadata.obs_size or -1), int(dataset.metadata.var_size or -1))
```
I think that should be the core of the changes you're making. You can remove `n_var` and `n_obs` then.
Or you can do
`dataset.metadata.shape = (int(dataset.metadata.obs_size), int(dataset.metadata.var_size))` like you've had it before and set `obs_size ` and `var` size defaults to -1 using the `MetadataElement `definitions.